include/continuable/detail/support/asio.hpp

```diff
@@ -0,0 +1,175 @@
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+**/
+
+#ifndef CONTINUABLE_DETAIL_ASIO_HPP_INCLUDED
+#define CONTINUABLE_DETAIL_ASIO_HPP_INCLUDED
+
+#if defined(ASIO_STANDALONE)
+#include <asio/async_result.hpp>
+#include <asio/error_code.hpp>
+#include <asio/version.hpp>
+
+#if (ASIO_VERSION / 100 % 1000) <= 12
+#define CTI_DETAIL_ASIO_HAS_NO_INTEGRATION
+#elif (ASIO_VERSION / 100 % 1000) <= 14
+#define CTI_DETAIL_ASIO_HAS_EXPLICIT_RET_TYPE_INTEGRATION
+#endif
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+#include <asio/system_error.hpp>
+#endif
+
+#define CTI_DETAIL_ASIO_NAMESPACE_BEGIN namespace asio {
+#define CTI_DETAIL_ASIO_NAMESPACE_END }
+#else
+#include <boost/asio/async_result.hpp>
+#include <boost/system/error_code.hpp>
+#include <boost/version.hpp>
+
+#if (BOOST_VERSION / 100 % 1000) <= 69
+#define CTI_DETAIL_ASIO_HAS_NO_INTEGRATION
+#elif (BOOST_VERSION / 100 % 1000) <= 71
+#define CTI_DETAIL_ASIO_HAS_EXPLICIT_RET_TYPE_INTEGRATION
+#endif
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+#include <boost/system/system_error.hpp>
+#endif
+
+#define CTI_DETAIL_ASIO_NAMESPACE_BEGIN                                        \
+  namespace boost {                                                            \
+  namespace asio {
+#define CTI_DETAIL_ASIO_NAMESPACE_END                                          \
+  }                                                                            \
+  }
+#endif
+
+#if defined(CTI_DETAIL_ASIO_HAS_NO_INTEGRATION)
+#error \
+  "First-class ASIO support for continuable requires the form of "\
+  "`async_result` with an `initiate` static member function, which was added " \
+  "in standalone ASIO 1.13.0 and Boost ASIO 1.70. Older versions can be " \
+  "integrated manually with `cti::promisify`."
+#endif
+
+#include <utility>
+
+namespace cti {
+namespace detail {
+namespace asio {
+
+#if defined(ASIO_STANDALONE)
+using error_code_t = ::asio::error_code;
+using error_cond_t = std::error_condition;
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+using exception_t = ::asio::system_error;
+#endif
+#else
+using error_code_t = ::boost::system::error_code;
+using error_cond_t = ::boost::system::error_condition;
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+using exception_t = ::boost::system::system_error;
+#endif
+#endif
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept {
+  promise.set_exception(std::make_exception_ptr(exception_t(std::move(e))));
+}
+#else
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept
+    ->decltype(promise.set_exception(std::move(e))) {
+  return promise.set_exception(std::move(e));
+}
+
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept
+    ->decltype(promise.set_exception(error_cond_t(e.value(), e.category()))) {
```

Comment:
Is ` ::boost::system::error_condition` (`error_cond_t`) in the case of boost implicitily convertible to `std::error_condition` (`exception_t`)?

Reply:
It is not--however I just looked at the documentation for Boost.System and

>On a C++11 compiler, Boost.System now provides implicit conversions from boost::system::error_category, error_code, and error_condition to their standard equivalents from <system_error>.

>This allows libraries to expose a C++11 interface and report errors via std::error_code even when using Boost.System, directly or through a dependency such as Boost.ASIO.

Honestly I was not aware of this--I know some Boost libraries become `typedef`s for their `std` equivalents when they get standardised, but I did not suspect that here (or any interop) because Boost.System is much superior to the `<system_error>` facilities. 

So this means if `exception_t` is `std::error_condition` and `error_code_t` is from Boost, we can do `exception_t(e.value(), e.category())` which may simplify the case handling.